### DISCUSSION TITLE
Csla 6 fluent configuration - should the DataPortal() method return CslaOptions?

### DISCUSSION THREAD

**Discussion by TheCakeMonster** [MarimerLLC/csla#2691] [Ideas]
Would it be better if the DataPortal() method returned CslaOptions?

I do not think what I am suggesting was the intention of this method. I think the whole point of it was to offer a more fluent and discoverable mechanism for performing configuration. However, this fluency comes at the cost of ease of use; because DataPortal() returns a subset of configuration, it is not possible to chain other top level configuration methods after it.

```csharp
// This compiles
services.AddCsla(options => options
.Security(cfg => cfg.PrincipalCacheMaxCacheSize(10000))
.DataPortal()
.UseLocalProxy()
);

// This does not compile
services.AddCsla(options => options
.DataPortal()
.UseLocalProxy()
.Security(cfg => cfg.PrincipalCacheMaxCacheSize(10000))
);
```
We already have the opportunity to write this using an action - a nested lambda - as follows:
```csharp
services.AddCsla(options => options
.Security(cfg => cfg.PrincipalCacheMaxCacheSize(10000))
.DataPortal(cfg => cfg.UseLocalProxy())
);
```
So does DataPortal() returning the subtype DataPortalClientOptions help or hinder configuration?

I think I've been getting myself confused about where certain configurations are because AddServerSideDataPortal() can be accessed both ways.

This would be a breaking change for existing users; however, since we've already changed the configuration setup in Csla 6 a fair bit it doesn't seem too unreasonable to do this - if it is considered sufficiently beneficial.

**Comment by swegele** [comment]
I also got hung up on what you mention.  It's OK, just a bit of a speed bump.
I do think that returning CslaOptions would make it more useable.  There is a bit of shifting shells of effort around; maybe it is just preference.  I guess I'm just saying I also struggled with what you mention above and expected it to work the way you suggest.

**Reply by rockfordlhotka** [reply]
Yes, there is inconsistency in the API that I don't like - I feel like I was in the middle of switching from one model to another and didn't really go through the code as thoroughly as necessary 😳
